### ISSUE TITLE
[ECR] [request]: add "sinceImagePulled" countType to ECR Lifecycle policy

### ISSUE THREAD

**Issue by mattmessinger** [aws/containers-roadmap#921]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
Add a new `sinceImagePulled` countType to ECR Lifecycle policy.

**Which service(s) is this request for?**
ECR

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
I would like to create an ECR Lifecycle policy that is based on when an image was last pulled. I can use such policy to infer that if an image has not been pulled in the last N months then it is not being used and I can safely delete it.

**Are you currently working around this issue?**
Right now we have to carefully track which images are still in use by our various build and deployed systems. This is error prone and leads to accidental deletion of images that are still being used.

**Comment by git-pchauhan** [comment]
👍 Would love to see this getting prioritized (and am kinda surprised it's not already there!). 👍

**Comment by wangyun1517** [comment]
Waiting for this!!!

**Comment by daniel-baptista-travcorp** [comment]
:+1: Yes please

**Comment by git-pchauhan** [comment]
Totally ➕ 1️⃣

**Comment by azamsiddiqi3791** [comment]
Yes please

**Comment by shashankvs01** [comment]
Please prioritize !!

**Comment by sbkg0002** [comment]
Any news on this?

**Comment by Cjcascade** [comment]
👍

**Comment by wangyun1517** [comment]
Please, our billing is crying!!!

**Comment by qihonggang** [comment]
waiting for this! please prioritize!

**Comment by sbkg0002** [comment]
Any update on this? `We're Working On It` since half a year :)

**Comment by snay2** [comment]
I opened a new issue to publish metrics for a similar usecase, in case it's useful to anyone here: https://github.com/aws/containers-roadmap/issues/1587

**Comment by wayne-folkes** [comment]
Adding my voice to the chorus on this one. My team is pushing multi-arch images. Because of this, images are shown as untagged. Having a policy that simply deletes untagged images would be dangerous as I have no way if it is safe to delete. If I knew the image has not been pulled in the last N days would give us some confidence that we are deleting unused resources.

**Comment by sbkg0002** [comment]
@arunsollet what happend? [The metrics seem to be there now](https://aws.amazon.com/about-aws/whats-new/2022/01/amazon-ecr-monitor-repository-pull-statistics/)!

**Comment by snay2** [comment]
@sbkg0002 I see a metric for `RepositoryPullCount` in the docs for ECR private (released in January 2022), but not one that describes how recently an image was pulled. Can you give more detail of what you're seeing?

**Comment by ivan-moto** [comment]
Hey, any updates on this?

**Comment by jdkealy** [comment]
Any updates on this ?

**Comment by ivanychev** [comment]
Any progress on this?

**Comment by maherrj** [comment]
100% second this. We have thousands of images across hundreds of repositories. We provide the service to our production consumers. We need to remove images but careful as to not cause an outage where an image is in use.

We had to develop a custom solution to tag the images based on CloudTrail events. Pretty horrible workaround.

**Comment by 24601** [comment]
> 100% second this. We have thousands of images across hundreds of repositories. We provide the service to our production consumers. We need to remove images but careful as to not cause an outage where an image is in use.
> 
> We had to develop a custom solution to tag the images based on CloudTrail events. Pretty horrible workaround.

Us, too. But expecting AWS to do anything that saves their customers money is, well, not something I am holding my breath for.

**Comment by jlbutler** [comment]
Hi @aviau 

> For us that wouldn't be an issue as we pull often. It covers many use cases so why not just release it while you build whatever more advanced feature you want to build?

For sure. I was just calling out a concern that I continue to have about leveraging this value on its own to indicate that an image is not in use. But as I indicated, we definitely are planning some work to improve the usefulness of the attribute (it currently tracks manifest pulls, which can make things confusing if you really want to know if the image was pulled including its layers), and we're looking at integrating this into LCP. We are still working on our roadmap, but we'll share more when there's something concrete.

> That won't work for many use cases because not everyone that uses ECR has deployments inside AWS.

Yep it gets a bit tricky. Like you said, not everyone uses ECR on AWS, and not everyone using ECR deploys on AWS. The work I referred to may not be ECR-specific, potentially working for any image digest. Part of that could be opt-in solutions (e.g. a Kubernetes controller you can install to report image use). I don't think we can serve all use cases, but we're doing some research around that now, and don't have any formal features planned with it yet.

We'll post back here when we have a more concrete timeline for this LCP request, thanks again!

**Comment by mgarber-ops** [comment]
I can see this being a nice to have but I'd be careful in situations where underlying EKS nodes are caching container images for respective workloads

**Comment by jobimrobinsantos-drizly** [comment]
This feature would be extremely handy for my organization. In particular, we would like to implement this type of lifecycle on our pull-through-cache repositories.

**Comment by prashil-g** [comment]
this is very important feature to have. any update if anyone is looking into it?

**Comment by volk1234** [comment]
Well, I wait this for years but I believe that features that really helps to save costs are not in priority at all :)

**Comment by volk1234** [comment]
@jlbutler 
Any updates about research you'd mentioned?

**Comment by seabyrn** [comment]
> Hi all. We are tracking a `lastRecordedPullTime` but have not yet done work to integrate it into LCP. One concern we have is that while it's one piece of data, it doesn't necessarily indicate an image is safe to expire if it hasn't been pulled in some amount of time.

Doesn't "sinceImagePushed" (which is available in LCPs) suffer from the same shortcoming?

**Comment by abhishekkundalia** [comment]
Any plans to push this or any alternative hacks to achieve this?

**Comment by blowfishpro** [comment]
> any alternative hacks to achieve this?

It's possible to get the last pull time via API calls and then explicitly delete those images.  This could be done e.g. by a lambda that runs periodically.

**Comment by nicc777** [comment]
Almost 4 years, for what I would have thought must be an obvious option to have. Can we perhaps get some kind of an update on this? @jlbutler ?

Also, the comment "_One concern we have is that while it's one piece of data, it doesn't necessarily indicate an image is safe to expire if it hasn't been pulled in some amount of time._" does not make sense, since I would argue `sinceImagePushed` leaves you with the same issue, if not even worse.

I am busy now implementing this on our side with a whole lot of logic just to refresh an image before it expires with the standard `sinceImagePushed` option. Having this option will hugely simplify our container management.

**Comment by sammcj** [comment]
I'm seeing clients getting impacted by not being able to check if an image is in use by ECS, came looking to see how to do this and found this issue. It is badly needed.

**Comment by sbkg0002** [comment]
Until someone writes a blogpost that goes viral, which describes the costs involved, nothing will happen apparently.

**Comment by Hronom** [comment]
AWS how about this proposal:

### Step 1
Introduce `last-pull-datetime` per image like in DockerHub:
https://docs.docker.com/docker-hub/api/latest/#tag/repositories/paths/~1v2~1namespaces~1%7Bnamespace%7D~1repositories~1%7Brepository%7D~1tags~1%7Btag%7D/get

![image](https://github.com/user-attachments/assets/4d01b3f1-e645-457b-bde3-bfbc8208f603)

### Step 2
Modify `Match criteria` and add there next options:
1. `Since image last pulled`, use days there same as for `Since image pushed`

### Step 3
Add to Lifecycle Policies the possibility to select `Sorting criteria` after `Match criteria` here:
![image](https://github.com/user-attachments/assets/9b9f24a1-9d6a-436d-9659-01d72c245764)

Inside `Sorting criteria` make several options:
1. `Unspecified` (Default)
3. `By image push date`
4. `By image last pull date`

Frontend logic:
1. If in `Match criteria` user selects `Since image pushed` - gray out `Sorting criteria` or not show them
2. If in `Match criteria` user selects `Since image last pulled` - gray out `Sorting criteria` or not show them
3. If in `Match criteria` user selects `Image count more then` - show `Sorting criteria`

Backend logic:
1. If in `Match criteria` user selects `Since image pushed` - do what you do right now
2. If in `Match criteria` user selects `Since image last pulled` - expire all images that have last pulled date above mentioned days
3. If in `Match criteria` user selects `Image count more then` - apply selected `Sorting criteria` first and then expire docker images that are beyond the mentioned limit for `Image count more then`

# Result
User be able to apply simple expiration by image pull date, based on `Since image last pulled` from `Match criteria`.

If users wanna have protection, they can leverage `Match criteria` in combination with `Sorting criteria` and set something like this: retain 10 of the most recent docker images.
This helps to avoid issue if you have docker images that are not frequently pulled.

**Comment by jammerful** [comment]
+1, this would be nice for us

**Comment by zcicala** [comment]
I am very surprised that this is not already a feature

**Comment by jarubert** [comment]
+1 waiting on this

**Comment by sbkg0002** [comment]
5 years and nothing happend. Thanks.

**Comment by tiago-vieira-sqills** [comment]
@tabern could you please give us some update on this? It seems this could be a bit game changer as the community is extremely asking for this?

Thanks in advance

**Comment by aviau** [comment]
Happy 5 years birthday to this feature request :)

**Comment by codedoga** [comment]
+1

**Comment by codedoga** [comment]
+1

**Comment by tomaecoward** [comment]
+1

**Comment by hypnopump** [comment]
+1

**Comment by irizzant** [comment]
+1

**Comment by cleivervx** [comment]
It is essential that we have this resource available.

**Comment by cgimeno** [comment]
+1

**Comment by ipleten** [comment]
Are there any plans to implement this?

**Comment by volphy** [comment]
lastRecordedPullTime is stored and available via AWS SDK/CLI and Management Console.

Will it be available in ECR Lifecycle Policy one day?

**Comment by JBATNCEK** [comment]
> Are there any plans to implement this?

On the AWS Events Channel, there's a video where they present this feature on their roadmap (starting at 38:40). The video is from December 5th, 2023, so it's a good question whether they'll actually implement it or not. 
https://www.youtube.com/watch?v=PHuKrcsAaDw

**Comment by nightraven-pb** [comment]
+++++1

**Comment by devinsmith911** [comment]
Looks like this functionality was released along with - https://aws.amazon.com/about-aws/whats-new/2025/11/amazon-ecr-archive-storage-class-container-images/
Can see `Days since last recorded pull time` in lifecycle criteria now

**Comment by nicc777** [comment]
> Looks like this functionality was released along with - https://aws.amazon.com/about-aws/whats-new/2025/11/amazon-ecr-archive-storage-class-container-images/ Can see `Days since last recorded pull time` in lifecycle criteria now

[I see it](https://imgur.com/a/uugtvsn) in `eu-central-1` as well. Great news!

EDIT: I see now as others have noted, you can not expire images. So, pretty useless.

**Comment by asimerel** [comment]
Yeah, but it won't expire images

<img width="899" height="317" alt="Image" src="https://github.com/user-attachments/assets/e38baa34-cc10-431e-ad4d-d0c55549e7f7" />

**Comment by zarko-a** [comment]
Not being able to expire the images is very disappointing. Almost makes this feature useless. Additionally, archiving has minimum charge of 90 day even if you manually deleted images that were auto archived.

**Comment by jeffkeller-einc** [comment]
I haven't seen it mentioned, but from my observations, Amazon Inspector scans will pull images and touch the "Last pulled at" timestamp. So this feature needs to ignore those pulls else it would be pretty useless when used in conjunction with Amazon Inspector.

**Comment by awsmjs** [comment]
Closing this ticket as it has been addressed as part of our new archival storage class. 

We've implemented the "sinceImagePulled" lifecycle policy paired with our new archival storage class to give customers an option to safely delete images. We've taken this approach after careful evaluations of numerous support cases where customers have unintentionally deleted critical images. This paired approach with archival will help customers to prevent accidental image deletion and give them a recovery window for their critical container images. 

- You can learn more about using this policy and tying it to eventual image deletion [here](https://docs.aws.amazon.com/AmazonECR/latest/userguide/lifecycle_policy_examples.html#lp_example_archive).
- To learn more about pull-time update exclusions, which allow you to specify IAM role ARNs that shouldn’t update image pull times when they pull images, you can read more [here](https://docs.aws.amazon.com/AmazonECR/latest/userguide/pull-time-update-exclusions.html).

We appreciate the community's feedback and remain committed to continuously improving our ECR lifecycle management capabilities.

**Comment by isker** [comment]
Hi @awsmjs. It's too bad that your other customers configure your software to delete images images that they don't want deleted, and then clog up your support queue complaining that they got what they asked for. The first behavior is an honest mistake, and the second behavior is simply dumb.

I'm not dumb. Please consider letting me configure your software to delete my images. Archival is not acceptable to me:
- It costs money. Because of the 90 day minimum, it costs more money than it costs to store the images I want deleted, up to the point where I want to delete them.
- I believe, but have not verified, that archived images still count against "images per repository" service quota, which my usage of ECR strains against. This quota is notionally configurable, but I have struggled to get it raised to the point where my use case, in the absence of some kind of image deletion on the basis of [`lastRecordedPullTime`](https://docs.aws.amazon.com/AmazonECR/latest/APIReference/API_ImageDetail.html#ECR-Type-ImageDetail-lastRecordedPullTime), can work.

To work around this situation, I've long had to develop, maintain, and operate a Lambda cronjob to implement the exact same behavior that you have now explicitly chosen not to expose natively through ECR lifecycles.

**Comment by gabrielbiasi** [comment]
It's laughable to see when you want to create **Repository lifecycle policy**, the AWS page does offer a example policy called "Expire on last pulled"
<img width="490" height="385" alt="Image" src="https://github.com/user-attachments/assets/5a619c7a-fa3d-4b4c-bdf7-f3b15398cc6f" />

The policy:
<img width="849" height="390" alt="Image" src="https://github.com/user-attachments/assets/83b05c91-f17c-4b28-b9b7-b03616e7fbcf" />

However, if you use this policy example, of course they block:
<img width="613" height="137" alt="Image" src="https://github.com/user-attachments/assets/eaaacc44-0e0e-4afe-8860-f1a577d108aa" />

It seems like it was expected for this feature to work, but they changed their minds ¯\\_(ツ)_/¯

**Comment by octane100** [comment]
Not being able to use **expire** with **sinceImagePulled** is particularly annoying for registries created by pull through cache (to mirror public upstream registries)

**Comment by mogopz** [comment]
@awsmjs can this please be reopened? The original issue description:
> I would like to create an ECR Lifecycle policy that is based on when an image was last pulled. I can use such policy to infer that if an image has not been pulled in the last N months then it is not being used and I can safely delete it.

This feature has been released but we can't delete images, we can only "archive" them.

It seems quite ridiculous to constrain the lifecycle since a few customers accidently delete images. Surely that happens all the time with other services that have lifecycle rules (S3, EBS, etc).

**Comment by matosf** [comment]
the [rationale given](https://github.com/aws/containers-roadmap/issues/921#issuecomment-3559226552) for not allowing the expiration/deletion of the images (i.e. only archive is allowed) is not aligned with many other AWS services with similar lifecycle rules, that do allow deletion of the underlying objects, with some of them being arguably even more business critical (e.g. RDS backups). 

the referenced risk to customers could be easily mitigated by making Archive the **default** Rule action, while still allowing Expire to be opted-in by anyone that wants to use (and thus understands what are the potential risks). as such, one can only conclude that AWS either think their customers are idiots or/and this is really all about keeping more $ in in their pockets. 

this paired with the pricing of this feature (i.e. only 30% cheaper and only over 150 TB) and the fact that it probably uses S3 Archive behind the curtains (which is 600% cheaper than S3 Standard), give us a good indication about the real motivation behind this (obviously bad) decision.

**Comment by dudicoco** [comment]
According to the updated examples page it seems that `sinceImagePulled` can be used with image expiry: https://docs.aws.amazon.com/AmazonECR/latest/userguide/lifecycle_policy_examples.html

**Comment by matosf** [comment]
@dudicoco unfortunately, as [this](https://github.com/aws/containers-roadmap/issues/921#issuecomment-3632625634) shows, if you try to configure it using the API it returns an error. 
Using the AWS Console is [also](https://github.com/aws/containers-roadmap/issues/921#issuecomment-3557621686) not possible (which I've also confirmed).

**Comment by dudicoco** [comment]
@matosf thanks for the info.
Too bad that AWS don't properly document the API on what is supported and what is not.

**Comment by ComfyDust** [comment]
<img width="969" height="723" alt="Image" src="https://github.com/user-attachments/assets/19a13623-c770-4473-81f3-72a497b146a6" />

<img width="852" height="716" alt="Image" src="https://github.com/user-attachments/assets/d9e0cd58-69c6-4bec-8253-e7b6b0dd1864" />

**Comment by matshch** [comment]
It looks like it does not help to collect grudge in this issue, better to upvote #2728.
